Add a guide on routing

jamesknelson opened the pull request:
This guide demonstrates how to implement a simple Router using the history library. It addresses the Routing item in #8060.

I've tried to keep it as brief as possible, there may be parts where I need to expand on the current content. Let me know/

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,241 @@
+    );
+  }
+}
+```
````
Please include CodePen links like in other docs.

jamesknelson replied:
I haven't found a way to make CodePen and similar tools work with routing examples. They disable access to `pushState` -- probably for good reason.

For the RRv4 and Junctions docs, we're using an in-memory history object and a "fake" url bar component. But this sounds a little heavy for the React website?

gaearon replied:
Ooh okay. Let's make sure the initial code snippet is copy-pasteable into Create React App without changes then. Like in RR4 docs.

jamesknelson replied:
Yep, have already tested this -- except moving `window.location` out of the component kind of breaks it.

gaearon replied:
Yea, okay let's keep it there.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+    if (location.pathname == '/news') {
+      content = <h1>News</h1>;
+    }
+    else if (location.pathname == '/profile') {
```
Style nit: please put `else` on the previous line.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+
+class App extends React.Component {
+  render() {
+    const location = window.location;
```
It's unfortunate we're reading a global from `render`. Generally this is a bad pattern. I would prefer if we did that once when we rendered `App`, and passed it as a prop.

gaearon replied:
This also means we need to include the code *rendering* the root component at least once. It's fine to omit it in later snippets.

jamesknelson replied:
Ok, will make this change.

The reason I had used the global-in-render method was that I use the same pattern with the `history` object in the later `<App>` and `<Link>` components. In order to avoid this pattern, I'd either need to use context to pass the `history` to the `<Link>` components, or we'd need to pass `history` explicitly via props. I'd like to hear your opinion on how to approach this.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+  );
+}
+
+class App extends React.Component {
```
Why is this a class? Since it doesn't use lifecycle or state, let's use a functional component.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+
+## HTML5 History
+
+The HTML5 [History API](https://developer.mozilla.org/en/docs/Web/API/History) makes it possible to update the browser's URL without causing a page reload. It also emits browser events when the user clicks *Back* or *Forward*. By using this API, you can write applications that respond faster and behave as expected when *Back* or *Forward* are clicked.
```
I'd like the next paragraph to be a higher level description of how apps using History API work. It can be as simple as explaining that History API exposes methods to rewrite the URL bar without actually changing the page, and events that fire when user presses Back and Forward. Imagine someone who's never seen client-side apps and write for that person.

jamesknelson replied:
Added as dot points, and removed the previous list which outlined the coming headings.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+
+Adding support for HTML5 History involves three steps:
+
+1. Exporting a history wrapper object
```
The reader doesn't even know the browser API at this point, so introducing "history wrapper object" sounds very complex. We could solve it by explaining the browser API first, and *then* noting its inconsistencies and introducing a library.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+- It will normalise events between browsers
+- It will emit navigation events caused by both the application and external stimuli
+
+While it is certainly possible to create your own wrapper, most React applications in the wild make use of the [history](https://github.com/mjackson/history) package.
```
Maybe change to "One popular implementation is [...]" rather than saying most people use it.

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,241 @@
+npm install history --save
+```
+
+As the History API is global, the wrapper will also need to be accessible everywhere. A good place to put it is in a `historyWrapper.js` file at the root of your source tree.
````
I want to avoid "wrapper" terminology in this doc, it just makes everything sound more complicated than it is. I would also avoid recommending any particular filename / location suggestions. In three months you'll see a cargo cult of history wrappers being placed in this location, and pointless discussions about it being a Best Practice. 😄

jamesknelson replied:
Will remove the filename suggestion.

Interesting point with `wrapper`, especially given that isn't just a wrapper as it provides extra functionality on top of the raw API.

The reason I named it as such is that just calling it `history` (which is what I've seen in the wild) may be confusing when the underlying API is `History`. What would you suggest the object be called?

gaearon replied:
I think it's okay since nobody actually uses History. :-P

gaearon replied:
Maybe let's just call it `browserHistory` since that's what history package calls it anyway. And call the file the same way.

jamesknelson replied:
I'm happy to name the variable `browserHistory`, although I would suggest against calling the file that as there are multiple types of `history`, and browser history is just one. Like you said, "Best Practices".

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,241 @@
+
+By default, HTML `<a>` elements will always cause the page to reload when clicked. If you'd like to update the URL using the History API, you'll need to cancel the default behaviour and then implement an alternative.
+
+While you could do this by adding an `onClick` handler to each individual `<a>` element, React applications often use a `Link` component. Instances of this component behave exactly like `<a>` elements, except that instead of refreshing the page, they will update the URL in-place using `historyWrapper.push()`:
```
It sounds a bit like `Link` component already exists. I would rephrase to say that since most links in the app share the same logic, people often create a component for them.

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,241 @@
+
+```js
+class Link extends React.Component {
+  handleClick = (event) => {
````
We don't use experimental syntax in the docs unless they're about that syntax in particular.
Please use the bind-in-constructor pattern.

ThisIsMissEm commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,243 @@
+      content = <h1>News</h1>;
+    } else if (location.pathname == '/profile') {
+      content = <h1>Profile</h1>;
+    } lse {
```
Else typo

thijsw replied:
Little typo in the word 'else'

JimFung commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,243 @@
+- It will normalise events between browsers
+- It will emit navigation events caused by both the application and external stimuli
+
+While it is certainly possible to create your own objcet to manage access to the History API, in this guide we'll use the popular [history](https://github.com/mjackson/history) package.
```
small typo here: Objcet instead of Object.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,243 @@
+
+  render() {
+    return (
+      <a {...this.props} onClick={this.handleClick}>
```
Minor nit: technically `{this.props.children}` is not required if you already spread props.
But I guess that's clearer.

jamesknelson commented:
I've renamed the `history` object to `browserHistory`, and moved `window.location` back into the `App` component for copy-pastability.

Is there anything I've missed?

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+---
+
+Routing is the process of choosing what to render based on the browser's current URL.
+
```
Let's make another intro paragraph to explain that React itself doesn't provide any built-in routing solutions, but you can use component lifecycle and state (link it to the doc) to hook it up to the browser APIs.

gaearon replied:
We also need to make it clear that React doesn't *require* you to implement routing. For example, if you *don't* write a single page app, you can use React without these techniques.

gaearon replied:
Also let's mention that in this guide, we'll build a simple routing solution from scratch, but there are popular community projects such as React Router.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+  const location = window.location;
+
+  let content;
+  if (location.pathname == '/news') {
```
Nit: please use `===` everywhere

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+Single page applications will often make use of the following two parts of the History API:
+
+- The [popstate](https://developer.mozilla.org/en-US/docs/Web/Events/popstate) browser event, which is emitted when *Back* or *Forward* are clicked
+- The [History.pushState()](https://developer.mozilla.org/en-US/docs/Web/API/History_API#The_pushState()_method) method, which allows the app to replace the contents of the URL bar without actually navigating
```
Maybe call it `window.history.pushState` to make it explicit we're talking about browser API.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+
+Single page applications will often make use of the following two parts of the History API:
+
+- The [popstate](https://developer.mozilla.org/en-US/docs/Web/Events/popstate) browser event, which is emitted when *Back* or *Forward* are clicked
```
Maybe say "the `popstate` browser event on the `window` object" to make it clear how to subscribe to it.

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,238 @@
+While it is certainly possible to create your own object to manage access to the History API, in this guide we'll use the popular [history](https://github.com/mjackson/history) package.
+
+```bash
+npm install history --save
````
Nit: let's put `--save` first like we do in other docs

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,238 @@
+npm install history --save
+```
+
+As the History API is global, our object will also need to be accessible everywhere.
````
Maybe add "For example, we can create a separate file and export the object from it, so that any other modules can later import it."

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,238 @@
+```js
+import createBrowserHistory from 'history/createBrowserHistory';
+
+const browserHistory = createBrowserHistory();
````
Let's add a default export?

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,238 @@
+
+```js
+// Get the current browserHistory.
+const location = history.location
````
Please use semicolons everywhere.

gaearon commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,238 @@
+This guide will use the following methods and properties from the `browserHistory` object:
+
+```js
+// Get the current browserHistory.
````
I think you meant "Get the current location". Perhaps it would be more useful to show what that object would look like.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+
+// Listen for changes to the current location.
+const unlisten = browserHistory.listen((location) => {
+  // location is an object like window.location
```
The reader may not know what `window.location` is, let's just show what it is in the comment?

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+    }
+    
+    const isNotLeftClick = event.button !== 0;
+    const isModified = !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
```
Let's add a comment explaining why this is important.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+    
+    // Update the browser URL and notify any listeners
+    // added via browserHistory.listen()
+    browserHistory.push(location);
```
Maybe highlight this line? You can see how other docs highlight JS with ```javascript{A,B,C} syntax where A,B,C are comma-separated line numbers.

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+  componentDidMount() {
+    // Subscribe after the component has mounted, in case the
+    // component is being rendered server-side and subscribing
+    // doesn't make sense
```
This comment will probably confuse many readers. Can you reword?

gaearon commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,238 @@
+    this.unsubscribe = browserHistory.listen(location => {
+      // Update the component state whenever the browser's
+      // location is updated
+      this.setState({ location })
```
Minor nit: semicolons and `({location})`, we use this style elsewhere in the docs.

gaearon commented:
Finally: it would be great if we could add this as a simple example somewhere.

jamesknelson commented:
Will go through these tasks on the weekend. Re: the example, do you mean as an example repository, or something else?

gaearon commented:
>Will go through these tasks on the weekend.

Thanks!

>Re: the example, do you mean as an example repository, or something else?

I don’t know. Maybe we can revive the `examples` folder with actually useful examples.

jamesknelson commented:
I've addressed the issues you mentioned. I'd appreciate any feedback on the fleshed out introduction and the updated description of the `browserHistory` object.

I've also added an example, but it feels a little out of place. Perhaps this would be better as a separate PR? Also, it took a bit to figure out how to get the examples directory to work. This could use some documentation, but probably belongs in a separate issue.

Finally, I've added a link to react-router per your suggestion. I understand your reasoning, though I'm hesitant to link readers to a library whose API will completely change at any time now. That said, I'm pretty biased.

mjackson commented on docs/docs/routing.md:
````diff
@@ -0,0 +1,261 @@
+You could do this by adding an `onClick` handler to each individual `<a>` element, but since most links within an app will share the same logic, people often create a component for them. Instances of this component behave exactly like `<a>` elements, except that instead of refreshing the page, they will update the URL in-place using `browserHistory.push()`:
+
+```js{47}
+class Link extends React.Component {
````
This `Link` class basically re-creates everything we're already doing in react-router. You can literally read through [our click handler](https://github.com/ReactTraining/react-router/blob/a757d7f2c56bfbc54833a82620a67f8ddf83e7e7/packages/react-router-dom/modules/Link.js#L34) and see all of the exact same pieces as you've got in your `handleClick`.

Why not just promote using the router? It's virtually the same code.

ryanflorence replied:
Bring in path parsing and you're 90% to react router v4. This feels weird to me.

jamesknelson replied:
The Link component here was based on a fairly old version of the one in react-router. Maybe we could add a comment mentioning this?

That said, the idea is to show people how Routing works. Once they understand, they'll have a better idea of whether it makes sense to use react-router for their project -- or possibly another router.

> Bring in path parsing and you're 90% to react router v4. 

But not having path parsing makes it so much easier for beginners to understand what is actually happening. We're also missing redirects, routes, switches, context, etc. As it stands, this is a long way from explaining 90% of react router v4.

ryanflorence replied:
Yeah, I still stand by 90% if you brought in path-to-regexp. Redirect, Switch, Route, are all pretty basic wrappers around history and path-to-regexp, the LOC come from propTypes.

> Maybe we could add a comment mentioning this?

You're actually supposed to respect the license of the code when you copy it. It's MIT.

Anyway, I'll bow out. Carry on.

dmitriid replied:
Basically, it all comes down to the purpose of the guide.

Is the guide supposed to explain react-router? Well, then "90%" (which still require "basic wrappers" such as Redirect, Switch, Route etc. etc. etc.) is perfectly ok.

However, the guide is supposed to be an explanation on how things can work/can be implemented in React. As such, it [must](https://www.ietf.org/rfc/rfc2119.txt) be as framework-agnostic as possible. It [should](https://www.ietf.org/rfc/rfc2119.txt) explain how a basic routing solution can be implemented. 

After that it should say:

> This is how you could implement a basic routing solution. There are also many routing libraries with many additional features. They can be either fully component-based ([react-router](https://github.com/ReactTraining/react-router)), designed for React ([monorouter](https://github.com/matthewwithanm/monorouter)) or framework-agnostic ([router5](http://router5.github.io))

mjackson replied:
> Maybe we could add a comment mentioning this?

The fact that you failed to mention that you copied that code from the router leads me to believe you're strongly biased against mentioning react-router at all.

dmitriid replied:
To close the whole topic revolving around hurt feelings by `react-router`, I propose simplifying the guide.

As we are building a basic routing solution first, we can remove all the nitty-gritty details of handing `onClick`. Start simple: do a `e.preventDefault()` and call some transition function unconditionally:

> ```js
>  handleClick(event) {
>     event.preventDefault();
>     // Instead of a normal URL, browserHistory.push() expects an object with
>     // `pathname` and `search` properties
>     const pathname = this.props.href.split('?')[0];
>     const search = this.props.href.slice(pathname.length);
>     const location = {pathname, search};
>     
>     // Update the browser URL and notify any listeners added
>     // via browserHistory.listen()
>    browserHistory.push(location);
>  }
> ```

This will be a very basic working router implementation. It's short, it's sweet, it works, it's perfect for a beginner.

After this is done, after other pieces like rendering proper components in `App` are explained, briefly explain other changes you can make to the `Link` component:

> To extend `Link` functionality, you may want to extend `handleClick` to correctly work in a variety of situations:
> - support for `onClick={...}` may be handled by `if(this.props.onClick) ...`
> - check if the user clicked left button, not right button: `const isNotLeftClick = event.button !== 0;`

And conclude with the paragraph I proposed [here](https://github.com/facebook/react/pull/8848#discussion_r105359987)

mjackson replied:
There are no "hurt feelings". [Our code is copyrighted](https://github.com/ReactTraining/react-router/blob/master/LICENSE.md). You can't just copy it and put it somewhere else as if it were your own.

jamesknelson replied:
@mjackson I learned a lot from your code and really grateful that you've made it available. I definitely did not intend to copy/paste without attribution.

The `Link` component does contain a lot of similar logic in the `onChange` handler, as it performs the same task -- but it is not a direct copy/paste. This is what I meant by "based on". But if you feel like it is a copyright violation, I will remove this PR and write a new `onChange` handler in my own library. Please let me know.

ryanflorence commented on docs/docs/routing.md:
```diff
@@ -0,0 +1,261 @@
+
+React itself does not have a built-in routing solution. But if you need one, component [state and lifecycle](/react/docs/state-and-lifecycle.html) are all you need to hook React up to the relevant browser APIs. This guide will take you through the process of building a simple Router using these tools.
+
+There are also a number of community projects, such as [react-router](https://github.com/ReactTraining/react-router), that can handle this task for you.
```
@dimitriid we haven't had a breaking change in the API for almost 2 years, history's props are no long spread into the route component props in v4, and named routes *prevent the ability to do dynamic routing and code splitting* (can also be implemented in like 10 lines of code on top)

dmitriid replied:
@ryanflorence 

[This issue](https://github.com/ReactTraining/react-router/issues/4424) has [three](https://github.com/ReactTraining/react-router/issues/4469) [other](https://github.com/ReactTraining/react-router/issues/4497) [issues](https://github.com/ReactTraining/react-router/issues/4519) referencing it. In every single one of them was closed because "that's the way it's going to be".

As for

> can also be implemented in like 10 lines of code on top

that's _exactly_ what you argued for in [the comment](https://github.com/ReactTraining/react-router/issues/4424#issuecomment-276742172) to the props issue. Moreover, that's basically the only argument you give over and over and over and over and over again:

"This is not needed, devs are wrong, you can always implement it yourself with 10 lines of code". Until, obviously, you figure out how to do it and suddenly it's ok, and makes its way into react-router code. 

However. Even with all that this is completely beyond the point.

The main point is: _this is not a react-router guide_. And React team should not endorse a single router solution even if it is named _react_-router. There are multiple other solutions. Some of them better. Some of them worse. Some using completely different approaches to react-router.

gaearon commented:
To be clear, the intended purpose of this guide is to explain routing is not something magical, but it's something you can do with history API and `setState`. The intention is not to show something too different from React Router, but to help people understand how routing could be implemented in React’s paradigm.

>As we are building a basic routing solution first, we can remove all the nitty-gritty details of handing onClick. Start simple: do a e.preventDefault() and call some transition function unconditionally:

Oh no, we can’t do this. People who care about accessibility will rightly accuse us of encouraging bad patterns in our own docs, which has happened in the past.

mjackson commented:
> The intention is not to show something too different from React Router, but to help people understand how routing could be implemented in React’s paradigm at all.

In that case I'd recommend just using hash URLs and using regular `<a>` tags instead of your own `<Link>`. Then implement a `hashchange` listener and a `switch` statement to choose which component to render and you're done.

```js
const HomePage = () => <div>Home page</div>
const AboutPage = () => <div>About page</div>
const NotFoundPage = () => <div>Not found</div>

class App extends React.Component {
  state = {
    url: this.getURL()
  }

  getURL() {
    return window.location.hash.substring(1)
  }

  handleHashChange = () => {
    this.setState({ url: this.getURL() })
  }

  componentDidMount() {
    window.addEventListener('hashchange', this.handleHashChange)
  }

  componentWillUnmount() {
    window.removeEventListener('hashchange', this.handleHashChange)
  }

  render() {
    const { url } = this.state

    let Child
    switch (url) {
      case '/':
        Child = HomePage
        break
      case '/about':
        Child = AboutPage
        break
      default:
        Child = NotFoundPage
    }

    return (
      <div>
        <h1>Welcome to the app!</h1>
        <nav>
          <ul>
            <li><a href="#/">Home</a></li>
            <li><a href="#/about">About</a></li>
          </ul>
        </nav>
        <Child/>
      </div>
    )
  }
}
```

Anything fancier than that and you may as well be using the router.

gaearon commented:
This sounds very reasonable to me. I agree we seem to have come to a dead end with this PR because the scope seems larger than I expected.

gaearon commented:
I think further arguments on this thread won't be very productive, and might best be held elsewhere.

The approach suggested in https://github.com/facebook/react/pull/8848#issuecomment-285901680 solves several problems brought up in this PR, and I'd take another PR implementing it.

gaearon commented:
To sum this up:

I'd like to express my gratitude to @jamesknelson who spent a lot of effort volunteering to write this guide when we called for contributions many months ago, and continuously changing it in response to my suggestions. I'm sad it didn't work out, but it's hard to know the scope beforehand. For example, the `fetch()` guide also got stuck once we realized the scope is too broad and doesn't really help React users as much. I feel like this is also what happened here, and I'm sorry for not realizing this sooner. Using browser history (and as a result needing a Link component) was my idea, and it was probably not appropriate for this kind of guide. 

I am also thankful to @ryanflorence and @mjackson for bringing up their concerns. I agree we shouldn't have copied that code from the router's Link implementation. I excused this in my head because it's hard to write a couple of early exits checking a few DOM event properties in different ways, and not writing them would break accessibility and/or usability.  But I should've made sure we recreated it from scratch in this PR instead of retaining similarity with RR implementation. 

Finally, it *was* the intention of this PR to link to React Router as a better implementation of that code. If you read the original issue that lead to this PR, it is explicitly written in https://github.com/facebook/react/issues/8060:

>Routing (no need to use React Router here, just show how to use setState with HTML5 history API **and point to RR as a component-based API for it once v4 is out**)

I promise I didn't edit that in today :wink: . So that was the intention all along, and we only didn't do it because RR4 was not released. (It was also the reason I didn't merge this PR since that was still the remaining item.)

To sum up, I'm sorry I mishandled this PR and didn't really do a good job of communicating what this is about to all interested parties. This is my fault. I think, however, that it would help all of us to assume good intentions in others.
